app/Router.py

```diff
@@ -51,12 +145,63 @@ class HostAndPathMatches(PathMatches):
 
     def __init__(self, host, path_pattern):
         super().__init__(path_pattern)
+        self.path_pattern = path_pattern
         self.host = host
+        self.match_cache = {}
 
     def match(self, request):
-        # Truncate the ".storyscriptapp.com" from "foo.asyncyapp.com"
-        if request.host[:-(Config.PRIMARY_DOMAIN_LEN + 1)] == self.host:
-            return super().match(request)
+
+        # Truncate the ".storyscriptapp.com" from "foo.asyncyapp.com" and
+        # ignore ports for local debugging
+        host = request.host.split(':')[0]
+
+        if host[:-(Config.PRIMARY_DOMAIN_LEN + 1)] == self.host.split(':')[0]:
+
+            # we know it's a wildcard match,
+            # so let's just go ahead and speed things up
+            if self.path_pattern is wild_matcher:
+                return {
+                    'path_args': [],
+                    'path_kwargs': {
+                        'wildcard': request.path.encode(),
+                        'endpath': None
+                    }
+                }
+
+            request_key = (self.host + request.path)
+
+            # below we override the super, and implement
+            # our own map based cache. This helps us
+            # save a few ms from caching
+            if request_key in self.match_cache:
+                return self.match_cache[request_key]
+
+            # This avoids the need for regex when the
+            # path_pattern is exactly the same as the
+            # request path.
+            safe_path = request.path.split('?')[0]
+            safe_pattern = str(self.path_pattern)
+
+            if safe_path in safe_pattern:
+                if safe_path == safe_pattern:
+                    return {}
+                elif safe_pattern.endswith('/') and \
+                        safe_pattern.startswith(safe_path) and \
+                        not safe_path.endswith('/') and \
+                        safe_pattern == (safe_path + '/'):
+                    self.match_cache[request_key] = {}
+
+                    return {}
+
+            value = super().match(request)
+
+            # we don't store anything for wildcard matched
+            # routes. We do this so we don't fill up the
+            # memory via malicious requests
+            if '?P<wildcard>' not in str(self.path_pattern):
```

Comment:
```suggestion
            if '?P<wildcard>' not in path_pattern:
```

That was already done ten lines above.